### PR TITLE
feat: sfn delete workflow (with prod token validation and messaging)

### PR DESCRIPTION
**Pull request by saikonen**
- additional changes on top of #1284 branch
- unable to open a PR targeting the fork branch, possibly simply not permitted.

Would prefer if original author would incorporate these into their own branch so #1284 can be merged. Otherwise pick one of the following strategies:
- this PR supersedes #1284
- merge #1284 in, crediting the original author. I'll Rebase this PR and merge afterwards.

### CODE REVIEW

File: metaflow/plugins/aws/step_functions/step_functions_cli.py [hunk 1]
```diff
@@ -545,3 +534,91 @@ def list_runs(
...
+
+
+def echo_token_instructions(obj, name, prev_user, cmd_name, cmd_description=None):
+    obj.echo(
+        "There is an existing version of *%s* on AWS Step "
+        "Functions which was deployed by the user "
+        "*%s*." % (name, prev_user)
+    )
+    obj.echo(
+        "To deploy a new version of this flow, you need to use "
```
Comment (savingoyal) [inline metaflow/plugins/aws/step_functions/step_functions_cli.py:613]: this wouldn't work for deletes. also, it is likely that we may have more commands that wouldn't allow us to combine the token instructions into one. Maybe it is indeed okay to have some repetition.

**Review by savingoyal — CHANGES_REQUESTED** [review]
few comments. I will test out the PR once I am back in office.

File: metaflow/plugins/aws/step_functions/step_functions_cli.py (hunk 1 above)
Comment (saikonen) [reply metaflow/plugins/aws/step_functions/step_functions_cli.py]: refactored this so token usage instructions can be passed to the validator as a function instead, keeping the `validate_token()` reusable for future commands as well (such as `terminate` etc. if we end up implementing those)

**Comment by savingoyal** [comment]
we can make this part of the next release train

File: metaflow/plugins/aws/step_functions/step_functions_cli.py [hunk 2]
```diff
@@ -545,3 +545,98 @@ def list_runs(
...
+            'See "Organizing Results" at docs.metaflow.org for more '
+            "information about production tokens."
+        )
+
+    validate_token(
+        obj.state_machine_name, obj.token_prefix, authorize, _token_instructions
+    )
+
+    obj.echo(
+        "Deleting Step Function *{name}*...".format(name=obj.state_machine_name),
```
Comment (savingoyal) [inline metaflow/plugins/aws/step_functions/step_functions_cli.py:583]: Minor nit - `AWS Step Functions state machine ...`

File: metaflow/plugins/aws/step_functions/step_functions_cli.py [hunk 3]
```diff
@@ -545,3 +545,98 @@ def list_runs(
...
+        bold=True,
+    )
+    schedule_deleted, sfn_deleted = StepFunctions.delete(obj.state_machine_name)
+
+    if schedule_deleted:
+        obj.echo(
+            "Deleting EventBridge rule *{name}* as well...".format(
```
Comment (savingoyal) [inline metaflow/plugins/aws/step_functions/step_functions_cli.py:590]: Minor nit - `Amazon EventBridge rule ...`

File: metaflow/plugins/aws/step_functions/step_functions_cli.py [hunk 4]
```diff
@@ -545,3 +545,98 @@ def list_runs(
...
+                name=obj.state_machine_name
+            ),
+            bold=True,
+        )
+    if sfn_deleted:
+        obj.echo(
+            "Deleting the Step Function may take a while. "
```
Comment (savingoyal) [inline metaflow/plugins/aws/step_functions/step_functions_cli.py:597]: `AWS Step Functions state machine ...`

Comment (savingoyal) [reply metaflow/plugins/aws/step_functions/step_functions_cli.py:597]: and below

File: metaflow/plugins/aws/step_functions/step_functions_cli.py [hunk 5]
```diff
@@ -545,3 +545,100 @@ def list_runs(
...
+            "Deleting the AWS Step Functions state machine may take a while. "
+            "Deploying the flow again to AWS Step Functions while the delete is in-flight will fail."
+        )
+        obj.echo(
+            "In-flight executions will not be affected. "
+            "If necessary, terminate them manually."
+        )
+
+
+def validate_token(name, token_prefix, authorize, instruction_fn=None):
```
Comment (savingoyal) [inline metaflow/plugins/aws/step_functions/step_functions_cli.py:608]: Can you add a small TODO that talks about unifying this method with `resolve_token` in the future (unless you prefer doing it as part of this PR)

**Review by savingoyal — CHANGES_REQUESTED** [review]
small nit - otherwise good to ship

**Review by savingoyal — APPROVED** [review]

**Pull request merged**